packages/hardhat-core/src/internal/core/providers/lazy-initialization.ts

```diff
@@ -40,6 +40,19 @@ export class LazyInitializationProviderAdapter implements EthereumProvider {
         this._initializingPromise = this._providerFactory();
       }
       this.provider = await this._initializingPromise;
+
+      // Copy any event emitter events before initialization over to the provider
+      const recordedEvents = this._emitter.eventNames();
+
+      for (const event of recordedEvents) {
+        const listeners = this._emitter.rawListeners(event) as Listener[];
+        for (const listener of listeners) {
+          this.provider.on(event, listener);
+          this._emitter.removeListener(event, listener);
+        }
+      }
+
+      this.provider.setMaxListeners(this._emitter.getMaxListeners());
```

Comment:
I know this is how it's done in the original code, but shouldn't we do it before the loop?